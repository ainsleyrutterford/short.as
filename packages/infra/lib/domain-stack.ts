import * as acm from "aws-cdk-lib/aws-certificatemanager";
import * as cdk from "aws-cdk-lib";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface DomainStackProps extends cdk.StackProps {}

/**
 * This stack creates:
 * - the Route53 HostedZone that will act as the DNS service for the short.as domain
 * - the SSL/TLS certificate for the domain
 * - the SSL/TLS certificate for the dev domain
 *
 * This should only be deployed to prod, and should only be deployed to the `us-east-1` region
 * since ACM certificates that are used with CloudFront must be in the `us-east-1` region.
 *
 * Documentation about using Route53 as the DNS service for a domain:
 * https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/migrate-dns-domain-inactive.html
 *
 * Documentation about accessing an ACM certificate cross stack/region:
 * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#cross-region-certificates
 */
export class DomainStack extends cdk.Stack {
  public hostedZone: route53.HostedZone;

  public certificate: acm.Certificate;
  public devCertificate: acm.Certificate;

  constructor(scope: Construct, id: string, props: DomainStackProps) {
    super(scope, id, props);

    // The name servers of this HostedZone are then manually copy/pasted into the short.as domain management
    // panel on the nic.as website
    this.hostedZone = new route53.HostedZone(this, "HostedZone", { zoneName: "short.as" });

    // Redirect www to the apex
    new route53.CnameRecord(this, "WwwCnameRecord", {
      recordName: "www",
      zone: this.hostedZone,
      domainName: this.hostedZone.zoneName,
    });

    this.certificate = new acm.Certificate(this, "Certificate", {
      domainName: "short.as",
      subjectAlternativeNames: ["www.short.as"],
      validation: acm.CertificateValidation.fromDns(this.hostedZone),
    });

    this.devCertificate = new acm.Certificate(this, "DevCertificate", {
      domainName: "dev.short.as",
      validation: acm.CertificateValidation.fromDns(this.hostedZone),
    });
  }
}
