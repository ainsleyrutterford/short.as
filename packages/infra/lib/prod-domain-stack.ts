import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cdk from 'aws-cdk-lib';
import * as route53 from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

interface ProdDomainStackProps extends cdk.StackProps {};

/**
 * This stack creates:
 * - the Route53 HostedZone that will act as the DNS service for the short.as domain
 * - the SSL/TLS certificate for the domain
 * 
 * This should only be deployed to prod, and should only be deployed to the `us-east-1` region
 * since ACM certificates that are used with CloudFront must be in the `us-east-1` region.
 * 
 * Documentation about using Route53 as the DNS service for a domain:
 * https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/migrate-dns-domain-inactive.html
 * 
 * Documentation about accessing an ACM certificate cross stack/region:
 * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#cross-region-certificates
 */
export class ProdDomainStack extends cdk.Stack {
  public hostedZone: route53.HostedZone;

  public certificate: acm.Certificate;

  constructor(scope: Construct, id: string, props: ProdDomainStackProps) {
    super(scope, id, props);

    this.hostedZone = new route53.HostedZone(this, 'HostedZone', { zoneName: 'short.as' });

    // TODO: set up redirects from www.short.as to short.as
    this.certificate = new acm.Certificate(this, 'Certificate', {
      domainName: 'short.as',
      subjectAlternativeNames: ['www.short.as'],
      validation: acm.CertificateValidation.fromDns(this.hostedZone)
    });
  }
}
