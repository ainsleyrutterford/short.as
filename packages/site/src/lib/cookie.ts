// https://stackoverflow.com/a/25490531
export const getCookieValue = (name: string) =>
  document.cookie.match("(^|;)\\s*" + name + "\\s*=\\s*([^;]+)")?.pop() || "";
