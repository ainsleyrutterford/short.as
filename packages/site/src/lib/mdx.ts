import { remark } from "remark";
import remarkFlexibleToc, { TocItem } from "remark-flexible-toc";
import { read } from "to-vfile";

/**
 * https://github.com/ipikuka/remark-flexible-toc
 */
export const getTocFromMdx = async (path: string) => {
  const toc: TocItem[] = [];

  await remark()
    // Extract the table of contents into `toc`
    .use(remarkFlexibleToc, { tocRef: toc })
    .process(await read(path));

  return toc;
};
