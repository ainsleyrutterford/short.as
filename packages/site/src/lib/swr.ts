export const USER_URLS_KEY = "/users/urls";

export const ANON_URLS_KEY = "/urls";
