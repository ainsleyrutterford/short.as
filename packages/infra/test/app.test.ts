import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { BackendStack } from '../lib/backend-stack';
import { WebsiteStack } from '../lib/website-stack';

// https://blog.bigbandsinger.dev/robust-cdk-snapshot-testing-with-snapshot-serializers
const s3KeyMatch = new RegExp(`[A-Za-z0-9]+\.zip`);
expect.addSnapshotSerializer({
  test: (val) => typeof val === 'string' && (val.match(s3KeyMatch) != null),
  print: (val) => {
    // Substitute both the bucket part and the asset zip part
    let sval = `${val}`;
    sval = sval.replace(s3KeyMatch, '[S3 KEY]');
    return `"${sval}"`;
  }
});

describe('Snapshot tests', () => {
  test('Backend stack matches snapshot', () => {
    /* Given */
    const app = new cdk.App();
    
    /* When */
    const stack = new BackendStack(app, 'TestTinyMuStack');
    const template = Template.fromStack(stack);
    
    /* Then */
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('Website stack matches snapshot', () => {
    /* Given */
    const app = new cdk.App();

    /* When */
    const backendStack = new BackendStack(app, 'TestTinyMuBackendStack');
    const websiteStack = new WebsiteStack(app, 'TestWebsiteStack', { httpApi: backendStack.httpApi });
    const template = Template.fromStack(websiteStack);

    /* Then */
    expect(template.toJSON()).toMatchSnapshot();
  });
});
