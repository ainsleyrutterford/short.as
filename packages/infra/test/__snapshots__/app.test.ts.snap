// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Backend stack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CountBucketsTable09F12BD0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "TableName": "TestBackendStack-CountBucketsTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "CreateShortUrlLambda0AED869B": {
      "DependsOn": [
        "CreateShortUrlLambdaServiceRoleDefaultPolicy64776861",
        "CreateShortUrlLambdaServiceRole009F038F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[S3 KEY]",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "COUNT_BUCKETS_TABLE_NAME": {
              "Ref": "CountBucketsTable09F12BD0",
            },
            "URLS_TABLE_NAME": {
              "Ref": "UrlsTable60368425",
            },
          },
        },
        "FunctionName": "TestBackendStack-CreateShortUrlLambda",
        "Handler": "index.createShortUrlHandler",
        "Role": {
          "Fn::GetAtt": [
            "CreateShortUrlLambdaServiceRole009F038F",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CreateShortUrlLambdaServiceRole009F038F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CreateShortUrlLambdaServiceRoleDefaultPolicy64776861": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CountBucketsTable09F12BD0",
                    "Arn",
                  ],
                },
                {
                  "Fn::GetAtt": [
                    "UrlsTable60368425",
                    "Arn",
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CreateShortUrlLambdaServiceRoleDefaultPolicy64776861",
        "Roles": [
          {
            "Ref": "CreateShortUrlLambdaServiceRole009F038F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetLongUrlLambdaC5112ECB": {
      "DependsOn": [
        "GetLongUrlLambdaServiceRoleDefaultPolicy0DF8ACD8",
        "GetLongUrlLambdaServiceRole025A1693",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[S3 KEY]",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "URLS_TABLE_NAME": {
              "Ref": "UrlsTable60368425",
            },
          },
        },
        "FunctionName": "TestBackendStack-GetLongUrlLambda",
        "Handler": "index.getLongUrlHandler",
        "Role": {
          "Fn::GetAtt": [
            "GetLongUrlLambdaServiceRole025A1693",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetLongUrlLambdaServiceRole025A1693": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetLongUrlLambdaServiceRoleDefaultPolicy0DF8ACD8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "UrlsTable60368425",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetLongUrlLambdaServiceRoleDefaultPolicy0DF8ACD8",
        "Roles": [
          {
            "Ref": "GetLongUrlLambdaServiceRole025A1693",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HttpAPI8D545486": {
      "Properties": {
        "CorsConfiguration": {
          "AllowMethods": [
            "GET",
            "POST",
          ],
          "AllowOrigins": [
            "https://short.as",
            "https://www.short.as",
            "https://dev.short.as",
            "http://localhost",
            "http://localhost:3000",
          ],
        },
        "Name": "TestBackendStack-HttpAPI",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "HttpAPIDefaultStage1BC7D78F": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpAPI8D545486",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "HttpAPIGETgetlongurlshortUrlIdA24C2A64": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpAPI8D545486",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /get-long-url/{shortUrlId}",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpAPIGETgetlongurlshortUrlIdGetLongUrlLambdaIntegrationF0B88EFF",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "HttpAPIGETgetlongurlshortUrlIdGetLongUrlLambdaIntegrationF0B88EFF": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpAPI8D545486",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "GetLongUrlLambdaC5112ECB",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpAPIGETgetlongurlshortUrlIdGetLongUrlLambdaIntegrationPermission117D995A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetLongUrlLambdaC5112ECB",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "HttpAPI8D545486",
              },
              "/*/*/get-long-url/{shortUrlId}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HttpAPIPOSTcreateshorturlCreateShortUrlLambdaIntegration86374E1F": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpAPI8D545486",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "CreateShortUrlLambda0AED869B",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpAPIPOSTcreateshorturlCreateShortUrlLambdaIntegrationPermissionF47EDEC2": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CreateShortUrlLambda0AED869B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "HttpAPI8D545486",
              },
              "/*/*/create-short-url",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HttpAPIPOSTcreateshorturlF335CED1": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpAPI8D545486",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /create-short-url",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpAPIPOSTcreateshorturlCreateShortUrlLambdaIntegration86374E1F",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "UrlsTable60368425": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "shortUrlId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "shortUrlId",
            "KeyType": "HASH",
          },
        ],
        "TableName": "TestBackendStack-UrlsTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Snapshot tests Website stack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:site/:0ebd2df5",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketDeploymentAwsCliLayerB018DCE1": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[S3 KEY]",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "BucketDeploymentCustomResourceCB7E2133": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketKeyPrefix": "site/",
        "DestinationBucketName": {
          "Ref": "Bucket83908E77",
        },
        "DistributionId": {
          "Ref": "Distribution830FAC52",
        },
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "[S3 KEY]",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketPolicyE9A3008A": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "OriginAccessIdentityDF1E3CAC",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "OriginAccessIdentityDF1E3CAC",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[S3 KEY]",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "BucketDeploymentAwsCliLayerB018DCE1",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Distribution830FAC52": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "dev.short.as",
          ],
          "CacheBehaviors": [
            {
              "CachePolicyId": "4135ea2d-6df8-44a3-9df3-4b5a84be39ad",
              "Compress": true,
              "FunctionAssociations": [
                {
                  "EventType": "viewer-request",
                  "FunctionARN": {
                    "Fn::GetAtt": [
                      "HtmlRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb738660F52",
                      "FunctionARN",
                    ],
                  },
                },
              ],
              "PathPattern": "/site*",
              "TargetOriginId": "TestWebsiteStackDistributionOrigin2E75A6923",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
          ],
          "CustomErrorResponses": [
            {
              "ErrorCode": 404,
              "ResponseCode": 404,
              "ResponsePagePath": "/site/404.html",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
              "PUT",
              "PATCH",
              "POST",
              "DELETE",
            ],
            "CachePolicyId": "4135ea2d-6df8-44a3-9df3-4b5a84be39ad",
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "RootRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7F8ADA6BC",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "TestWebsiteStackDistributionOrigin1F299B5C8",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Fn::Join": [
                          "",
                          [
                            "https://",
                            {
                              "Fn::ImportValue": "TestBackendStack:ExportsOutputRefHttpAPI8D545486FD78B06F",
                            },
                            ".execute-api.",
                            {
                              "Ref": "AWS::Region",
                            },
                            ".",
                            {
                              "Ref": "AWS::URLSuffix",
                            },
                            "/",
                          ],
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "TestWebsiteStackDistributionOrigin1F299B5C8",
              "OriginPath": "/get-long-url",
            },
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "Bucket83908E77",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestWebsiteStackDistributionOrigin2E75A6923",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "OriginAccessIdentityDF1E3CAC",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "HtmlRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb738660F52": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "/**
 * Redirects the following:
 * - https://short.as/site -> https://short.as/site/index.html
 * - https://short.as/site/folder/ -> https://short.as/site/folder/index.html
 * - https://short.as/site/page -> https://short.as/site/page.html
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
async function handler(event) {
  const request = event.request;

  if (request.uri === "/site") {
    request.uri += "/index.html";
  } else if (request.uri.endsWith("/")) {
    request.uri += "index.html";
  } else if (!request.uri.includes(".")) {
    request.uri += ".html";
  }

  return request;
}
",
        "FunctionConfig": {
          "Comment": "HtmlRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7",
          "Runtime": "cloudfront-js-2.0",
        },
        "Name": "HtmlRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7",
      },
      "Type": "AWS::CloudFront::Function",
    },
    "OriginAccessIdentityDF1E3CAC": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "RootRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7F8ADA6BC": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "/**
 * Redirects the following:
 * - https://short.as/ -> https://short.as/site/ so that CloudFront redirects to the S3 website
 * - https://short.as/aaaaaaa/ -> https://short.as/aaaaaaa so that CloudFront calls the \`get-long-url\` API correctly
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
async function handler(event) {
  const request = event.request;
  const host = request.headers.host.value;
  const newUrl = \`https://\${host}/site/\`;

  if (request.uri === "/") {
    return {
      statusCode: 302,
      statusDescription: "Found",
      headers: { location: { value: newUrl } },
    };
  } else if (request.uri.endsWith("/")) {
    request.uri = request.uri.slice(0, -1);
  }

  return request;
}
",
        "FunctionConfig": {
          "Comment": "RootRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7",
          "Runtime": "cloudfront-js-2.0",
        },
        "Name": "RootRedirectFunctionc8ac49a71b70e8f94e35d7058e80417b45c0334fb7",
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
